bgpd/bgp_route.c

```diff
@@ -2393,12 +2424,27 @@ void bgp_best_selection(struct bgp *bgp, struct bgp_dest *dest,
 			bgp_path_info_unset_flag(dest, pi, BGP_PATH_DMED_CHECK);
 			if (debug)
 				zlog_debug("%s: pi %p dmed", __func__, pi);
+
+			if (pi == old_select)
+				old_select_valid = false;
 			continue;
 		}
 
 		bgp_path_info_unset_flag(dest, pi, BGP_PATH_DMED_CHECK);
 
 		reason = dest->reason;
+
+		/* Compare paths till the old selected path is found
+		 * in the list
+		 */
+		if ((old_select_valid) && (pi == old_select->next))
```

Comment:
Consider this we have these 3 bgp path info's

A -> B -> C

best path selection has been run in the past and we are stable.  A has been moved to the front of the list.

A new update for A comes in that changes the attributes on A.  When we run best path selection over this C would be choosen prior to this change, but with this change I believe We would never consider the A -vs- C to see who would win, as that I do not believe we unset BGP_PATH_SELECTED

I would like a test case that covers this.

See the above comment about when we first set old_select_valid.  I believe this would fix this issue.

Reply:
Will  look at the different scenarios as mentioned in the comments,   the test case errors are probably due to change in the order of paths in the list